pkg/subscription/subscriptionmanager.go

```diff
@@ -0,0 +1,230 @@
+package subscription
+
+import (
+	"sync"
+
+	"github.com/Axway/agents-mulesoft/pkg/anypoint"
+
+	"github.com/Axway/agent-sdk/pkg/apic"
+	"github.com/Axway/agent-sdk/pkg/apic/apiserver/models/management/v1alpha1"
+	"github.com/sirupsen/logrus"
+)
+
+type SchemaHandler interface {
+	GetSubscriptionSchemaName(pd PolicyDetail) string
+	RegisterNewSchema(schemaConstructor SchemaConstructor, apc anypoint.Client)
+}
+
+// SchemaConstructor -
+type SchemaConstructor func(client anypoint.Client) Handler
+
+var constructors []SchemaConstructor
+
+func Register(constructor SchemaConstructor) {
+	constructors = append(constructors, constructor)
+}
+
+// ConsumerInstanceGetter gets a consumer instance by id.
+type ConsumerInstanceGetter interface {
+	GetConsumerInstanceByID(id string) (*v1alpha1.ConsumerInstance, error)
+}
+
+// SubscriptionsGetter gets the all the subscriptions in any of the states for the catalog item with id
+type SubscriptionsGetter interface {
+	GetSubscriptionsForCatalogItem(states []string, id string) ([]apic.CentralSubscription, error)
+}
+
+type Handler interface {
+	Schema() apic.SubscriptionSchema
+	Name() string
+	IsApplicable(policyDetail PolicyDetail) bool
+	Subscribe(log logrus.FieldLogger, subs apic.Subscription) error
+	Unsubscribe(log logrus.FieldLogger, subs apic.Subscription)
+}
+
+type PolicyDetail struct {
+	Policy     string
+	IsSlaBased bool
+	APIId      string
+}
+
+// Manager handles the subscription aspects
+type Manager struct {
+	log      logrus.FieldLogger
+	handlers map[string]Handler
+	cig      ConsumerInstanceGetter
+	sg       SubscriptionsGetter
+	dg       *duplicateGuard
+}
+
+type duplicateGuard struct {
+	cache map[string]interface{}
+	lock  *sync.Mutex
+}
+
+func New(log logrus.FieldLogger,
+	cig ConsumerInstanceGetter,
+	sg SubscriptionsGetter,
+	apc anypoint.Client,
+) *Manager {
+	handlers := make(map[string]Handler, len(constructors))
+
+	for _, c := range constructors {
```

Comment:
Yes, at this time clientId's init() calls the Register function to register its handler for client-id-enforcement auth policy. This can be extended to other types of auth in the future.